packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts

```diff
@@ -0,0 +1,601 @@
+ * @param messages messages to send
+ * @returns promise with an array of message ids
+ */
+async function startAndSendMultipleMessages(
+  messages: Array<{}>
+): Promise<string[]> {
+  const processingTask = processor.collectAndProcess();
+  const allMessages: Array<Promise<string>> = messages.map(message => {
+    const messageId = mockSubscription.sendMockMessage(asBuffer(message));
+    return processingTask.then(() => {
+      return messageId;
+    });
+  });
+  return Promise.all(allMessages);
+}
+
+describe('Cloud Logs Processor', () => {
+  describe('collectAndProcess()', () => {
+    beforeEach(() => {
+      mockFirestore = new MockFirestore({
+        Bot_Execution: {},
+        Error: {},
+        Trigger: {},
+        Action: {},
+        Action_Type: {},
+        GitHub_Repository: {},
+        GitHub_Object: {},
+      });
+      mockSubscription = new MockSubscription(
+        new PubSub(),
+        'mock-subscription'
+      );
+    });
+
+    describe('correctly formed incoming log messages', () => {
+      beforeEach(() => {
+        processor = new CloudLogsProcessor({
+          firestore: mockFirestore,
+          subscription: mockSubscription,
+        });
+      });
+
+      describe('execution start and execution end logs', () => {
+        // what the execution record would look like after receiving just the start log
+        const executionRecordStart = {
+          document: {
+            '4ww4alqs7ikq': {
+              execution_id: '4ww4alqs7ikq',
+              bot_id: 'merge_on_green',
+              start_time: 1595536893000,
+              logs_url:
+                'https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22' +
+                'cloud_function%22%0Alabels.%22execution_id%22%3D%224ww4alqs7ikq%22;' +
+                'timeRange=2020-07-23T20:41:33.701320846Z%2F22020-07-23T20:41:33.701320846Z;' +
+                'summaryFields=:true:32:beginning?project=repo-automation-bots',
+            },
+          },
+          collectionName: 'Bot_Execution',
+        };
+
+        // what the execution record would look like after receiving just the end log
+        const executionRecordEnd = {
+          document: {
+            '4ww4alqs7ikq': {
+              execution_id: '4ww4q2vqvkl1',
+              bot_id: 'auto_label',
+              end_time: 1595536887000,
+            },
+          },
+          collectionName: 'Bot_Execution',
+        };
+
+        // what the execution record would look like after receiving start + end logs
+        const executionRecordBoth = {
+          document: {
+            ...executionRecordStart.document,
+            ...executionRecordEnd.document,
+          },
+          collectionName: 'Bot_Execution',
+        };
+
+        describe('when no execution record exists', () => {
+          it('creates a new execution record and stores execution start logs', () => {
+            return testValidMessage(VALID_MESSAGES.execution_start, [
```

Comment:
This method returns a promise - do we need to await all of these to ensure the assertion runs/completes?

Reply:
If you return a promise mocha automatically waits for it's completion. It's equivalent to doing `await` and then `done()`

Reply:
Gotcha, TIL.